### PR TITLE
Refactor Outfit::Load() and support `add` and `remove` syntax

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #1764 

## Feature Details
This perhaps got a little out of hand.

Instead of doing `child.Token(0) == "[ reverse | steering] flare [ sprite | sound]"` over and over and over again, use a few maps which hold the `key`s and pointers to the members of the `Outfit` object that they correspond to, so we can just do a few `find`s and see if we find it and then do whatever to the thing at that iterator.
In the case of these properties, this means that adding additional similar members will not require updating multiple if statements, just add it to the appropriate map at the top of `Outfit::Load()`.

`licenses` is handled first because that node may or may not have a value and still be valid.

`remove <attribute>` with no additional tokens is handled next.
This case clears the collection (if it corresponds to a sprite, sound, or effect), clears the string (if it is one of the string members), or sets the value to 0, if it is any numerical attribute.

`remove <attribute> <value>`
In the case of sprites, sounds, and effects, removes the item with the name given in the value.
Has no meaning and will print an error if a different attribute is given.

`[add] <attribute> <value>`
Without `add`, will continue to operate as it currently does, overwriting numerical attribute values, adding additional sounds, and effects, overwriting sprites (flare, flotsam, or thumbnail), overwriting ammo, overwriting strings.
(This is not quite as it currently works, currently, additional `description` tokens append to the existing description, this will still happen within a single outfit definition, but plugins adding to the description of an outfit will now need to use `add` on their first `description` node to avoid overwriting the existing description, if that is the intention. The issue did ask for a way to overwrite the existing description.)
With `add`, for numerical attributes, the value will be added to the existing value, and, for sounds and effects, they may specify an additional token indicating how many of that sound or effect to add, this will be truncated to an integer.
Not all properties support `add`, sprites and ammo will print an error if `add` is used with them.

`weapon` is unaffected (except that you now can't have trailing tokens on the weapon node itself. You shouldn't anyway; they don't do anything.)

Good luck to whoever updates the wiki with this.

<summary>The old description.</summary>

<details>
Allows `remove`ing outfit attributes.
`remove` can either be used to clear certain collections (sprites, sounds, set attributes to 0) when it is followed by a single token. When followed by two or more, it will only remove a specific item from those collections.
Also add `add` for the attributes where it makes sense and the default behaviour isn't to add to the existing value.
<details><summary>The full list of things that can be removed:</summary>

`flare sprite` (also, `reverse` and `steering` versions)
`flare sound` (also, `reverse` and `steering` versions)
`afterburner effect`
`jump effect`
`hyperdrive sound`
`hyperdrive in sound`
`hyperdrive out sound`
`jump sound`
`jump in sound`
`jump out sound`
`description`
`licenses`

Additionally, all other attributes stored in the `attributes` collection can be `remove`ed, erasing that value (sets it to zero).
</details>

<details><summary>Attributes that can be added to:</summary>
`cost`
`mass`
`jump range` (still checks that the added value doesn't make it negative)
All attributes stored in `attributes` (so, any other attribute).
</details>

I haven't touched anything inside the `weapon` node here.
</details>

## Usage Examples
```
remove licenses Navy
  "Wanderer Outfits"
```

## Testing Done
It built and launched, will test further soon.

I have made the game print as much outfit data as I can get from it and compared the output with this refactor to that without and the results are identical.

## Performance Impact
N/A, I hope.

### CODE REVIEW

File: source/Outfit.cpp [hunk 1]
```diff
@@ -195,7 +195,90 @@ void Outfit::Load(const DataNode &node)
...
+			else if(key == "hyperdrive out sound")
+				hyperOutSounds.clear();
+			else if(key == "jump sound")
+				jumpSounds.clear();
+			else if(key == "jump in sound")
+				jumpInSounds.clear();
+			else if(key == "jump out sound")
+				jumpOutSounds.clear();
+			else if(key == "description")
+				description = "";
```
Comment (Hurleveur) [inline source/Outfit.cpp:230]:
```suggestion
				description.clear();
```

File: source/Outfit.cpp [hunk 2]
```diff
@@ -195,7 +195,90 @@ void Outfit::Load(const DataNode &node)
...
+			else if(child.Token(0) == "hyperdrive out sound" && child.Size() >= 2)
+				hyperOutSounds.erase(Audio::Get(child.Token(1)));
+			else if(child.Token(0) == "jump sound" && child.Size() >= 2)
+				jumpSounds.erase(Audio::Get(child.Token(1)));
+			else if(child.Token(0) == "jump in sound" && child.Size() >= 2)
+				jumpInSounds.erase(Audio::Get(child.Token(1)));
+			else if(child.Token(0) == "jump out sound" && child.Size() >= 2)
+				jumpOutSounds.erase(Audio::Get(child.Token(1)));
+			else if(key == "description")
+				description = "";
```
Comment (Hurleveur) [inline source/Outfit.cpp:260]:
```suggestion
				description.clear();
```

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Small things I noticed
By the way, should we not be adding support for the add key at the same time? We don't have to but that way it would allow full event support for outfits.

also I'm wondering, why don't we use switch cases instead of huuuuge else ifs

**Comment by Hurleveur** [comment]
The checks themselves warn about the &string
Also if you want inspiration on how to do this (with add too), it was recently done in Government.cpp

**Comment by warp-core** [comment]
The default behaviour is to `add` for all the attributes I've made `remove` lines for.

I suppose we could add `add` options for the currently overridden attributes, but it's not really meaningful for all of them. You can't add a thumbnail sprite to the existing one.

For each remaining attribute:

`cetegory` and `plural name`: `add` would just concatenate to the ends of the existing strings. Maybe if you wanted to move outfits into new, similarly names categories, without needing to know the original category? (`Systems2`, `Engines2`, etc). Can't imagine much use for the plural name.

`flotsam sprite` and `thumbnail` completely meaningless.

`cost`, `mass`, and `jump range`. I suppose it might be meaningful to `add` to these values.

`ammo` modifying this would just break stuff.

Any other attributes: I suppose `add` or `remove` are no more likely to break stuff than overriding these values already is.

**Comment by Hurleveur** [comment]
Add makes sense for description and license too, I think

**Comment by warp-core** [comment]
The existing behaviour for descriptions and licenses is to add. Not being able to remove the existing description, only being able to add to it with plugins was the reason for the issue this resolves.

**Comment by Hurleveur** [comment]
Yeah and the way to resolve it is to make add only happen with the add keyword, and default behaviour override the description.
That way when redefining an outfit you don't have to spam remove for every field.

**Comment by warp-core** [comment]
Applying that here would change the behaviour of existing syntax. While the syntax for outfits is a bit of a mess, my understanding is that changing the result of existing syntax is not good. Granted, changes to outfits can't be made in events so this would only affect plugins, not saves, but stil.

**Comment by warp-core** [comment]
Ok, added `add` for `cost`, `mass`, `jump range`, and mapped attributes, and allowed `clear`ing mapped attributes. `remove` would be the same as `clear` for mapped attributes so nothing happens there.

**Comment by warp-core** [comment]
All the other attributes are what I've referred to as 'mapped` (nothing particularly special is done when loading them, they are just shoved into a Dictionary object).

This still only supports the `add`, `remove`, and `clear` tokens in `Outfit::Load()`, which is only accessed when loading game data. I don't think it is currently possible for anything inside an event to be passed to `Outfit::Load()` so events still can't make use of this feature, it only allows plugins to modify outfits with more flexibility.

I think adding event support for outfits probably ought to be a separate PR, though this one does improve the options available if it is added (note that, if event support were added without the changes in this PR, the event would still be able to overwrite most outfit attributes the same way a plugin can).

**Comment by Zitchas** [comment]
Oh. Well, thanks for the explanation. I guess I'll contain my excitement until we get to intentional events-can-modify-definitions capabilities, then.

File: source/Outfit.cpp [hunk 3]
```diff
@@ -195,7 +195,104 @@ void Outfit::Load(const DataNode &node)
 
 	for(const DataNode &child : node)
 	{
-		if(child.Token(0) == "category" && child.Size() >= 2)
+		if(child.Token(0) == "clear" && child.Size() >= 2)
```
Comment (quyykk) [inline source/Outfit.cpp:198]: `"clear"` is deprecated syntax, no need to add support for it.

Comment (Hurleveur) [reply source/Outfit.cpp:198]: yeah but that said it should be replaced by "remove" without specifying what to remove
I kinda like the idea to be able to remove just one thing

Comment (warp-core) [reply source/Outfit.cpp]: `remove` followed by one token does what `clear` was doing.

**Comment by Amazinite** [comment]
By request of warp.

**Comment by warp-core** [comment]
Closing to appease the PR number.

**Comment by warp-core** [comment]
I finished it, maybe!

**Review by Hurleveur — APPROVED** [review]
Can't find anything wrong with it code wise, it's more proper than before in fact.
Just wondering, does this mean that variants will be able to define custom licenses?

**Comment by warp-core** [comment]
Part of this can be verified using the `--outfits --all` command, which prints every attribute of every outfit.
Doing this before and after this PR produces the same result (compared using git) so we can be confident that the attributes are all still loaded as they were before.

**Comment by warp-core** [comment]
By comparing the output of `--outfits --other` between the following two branches, I have verified that the other non-weapon properties of outfits are all loaded correctly.

https://github.com/warp-core/endless-sky/tree/Log-other-outfit-data-without-Outfit-Load-refactor
https://github.com/warp-core/endless-sky/tree/Log-other-outfit-data

Additionally, I have used the `--weapons` command to verify that all weapon data is still loaded the same.

**Comment by TomGoodIdea** [comment]
plz resolve conflicts

**Comment by tibetiroka** [comment]
Closing as orphaned. Feel free to reopen the PR once the conflicts and review comments are resolved.

**Pull request closed without merging**